shared-module/gamepad/__init__.c

```diff
@@ -34,25 +34,49 @@
 
 
 void gamepad_tick(void) {
+    static uint8_t last = 0;
+    uint8_t current = 0;
+    uint8_t bit = 1;
+    digitalio_digitalinout_obj_t* data_pin;
+    digitalio_digitalinout_obj_t* clock_pin;
+    digitalio_digitalinout_obj_t* latch_pin;
+
     gamepad_obj_t* gamepad_singleton = MP_STATE_VM(gamepad_singleton);
     if (!gamepad_singleton) {
         return;
     }
-    uint8_t gamepad_current = 0;
-    uint8_t bit = 1;
-    for (int i = 0; i < 8; ++i) {
-        digitalio_digitalinout_obj_t* pin = gamepad_singleton->pins[i];
-        if (!pin) {
-            break;
+    if (gamepad_singleton->pins[0]) {
+        // buttons connected directly to pins
+        for (int i = 0; i < 8; ++i) {
+            digitalio_digitalinout_obj_t* pin = gamepad_singleton->pins[i];
+            if (!pin) {
+                break;
+            }
+            if (common_hal_digitalio_digitalinout_get_value(pin)) {
+                current |= bit;
+            }
+            bit <<= 1;
         }
-        if (common_hal_digitalio_digitalinout_get_value(pin)) {
-            gamepad_current |= bit;
+        current ^= gamepad_singleton->pulls;
+    } else {
+        // buttons connected to a shift register
+        data_pin = gamepad_singleton->pins[1];
+        clock_pin = gamepad_singleton->pins[2];
+        latch_pin = gamepad_singleton->pins[3];
```

Comment:
This is a bit ugly. Instead, how about casting the gamepad singleton to gamepadshift once we know what it is?

Reply:
Not sure how that would be done. Do you have an example?

Reply:
Ah, I thought you already have a separate struct for gamepad_shift_obj. I'd recommend adding a separate one. It will take less memory than direct gamepad because it holds fewer pins.

Reply:
I think I understand now. You mean the singleton can be any python object, so I can store either one. Thanks.

Reply:
Yup! Exactly!